File: Python/Design/L_DesignBrowserHistory.py
class BrowserHistory:
    
    def __init__(self, homepage: str):
        self.history = [homepage]
        self.curr = 0
        
    def visit(self, url: str) -> None:
        self.curr += 1
        while self.curr < len(self.history):
            self.history.pop()

        self.history.append(url)
        
    def back(self, steps: int) -> str:
        self.curr = max(0, self.curr-steps)
        return self.history[self.curr]
      
    def forward(self, steps: int) -> str:
        self.curr = min(len(self.history)-1, self.curr+steps)
        return self.history[self.curr]
        


# * Linked List version
class ListNode:
    def __init__(self, val, next = None, prev = None):
        self.val = val
        self.next = next
        self.prev = prev

class BrowserHistory1:

    def __init__(self, homepage: str):
        self.curr = ListNode(homepage)
        
    def visit(self, url: str) -> None:
        self.curr.next = ListNode(url, None, self.curr)
        self.curr = self.curr.next
        
    def back(self, steps: int) -> str:
        while self.curr.prev and steps > 0:
            self.curr = self.curr.prev
            steps -= 1
        return self.curr.val
      
    def forward(self, steps: int) -> str:
        while self.curr.next and steps > 0:
            self.curr = self.curr.next
            steps -= 1
        return self.curr.val
        

File: cpp/Linked List/L_designBrowserHistory.cpp
#include<bits/stdc++.h>
using namespace std;

struct listnode {
    string val;
    listnode *next;
    listnode *prev;
    listnode() : val(""), next(nullptr), prev(nullptr) {}
    listnode(string x) : val(x), next(NULL), prev(NULL) {}
};

class BrowserHistory {
public:
    listnode *curr;
    BrowserHistory(string homepage) {
        curr = new listnode(homepage);

    }
    
    void visit(string url) {
        curr->next = nullptr;
        listnode *newPage = new listnode(url);
        curr->next = newPage;
        newPage->prev = curr;
        curr = newPage;
    }
    
    string back(int steps) {
        while(curr->prev and steps--) {
            curr = curr->prev;
        }
        return curr->val;
    }
    
    string forward(int steps) {
        while(curr->next and steps--) {
            curr = curr->next;
        }
        return curr->val;
    }
};

